File: lib/helpers/matrix.dart
import "dart:collection" show IterableMixin;

import "range.dart";

class Fraction {
	final int nom, denom;
	final bool negative;
	const Fraction._ (this.nom, this.denom, this.negative);

	factory Fraction (num nom, num denom) {
		assert (nom is int && denom is int, "Fraction received double");
		if (denom == 0) throw IntegerDivisionByZeroException();
		final bool negative = nom >= 0 != denom >= 0;
		nom = nom.abs();
		denom = denom.abs();
		for (final int n in range (1, start: nom, step: -1)) {
			if (nom % n == 0 && denom % n == 0) {
				nom ~/= n;  // integer division
				denom ~/= n;
			}
		}

		return Fraction._ (nom, denom, negative);
	}

	double get toNum => nom / denom;
}

class Slice {
	final int start, end;
	const Slice (this.start, this.end);
}

class Pivot {
	final int index;
	final double value;
	Pivot (Pair<double> pair) : index = pair.index, value = pair.value;
	Pivot.empty() : index = null, value = null;
}

int lcm (List<Fraction> fractions) {
	int maxDenom;
	Set<int> denoms = Set<int>();

	// find the largest denominator
	for (final Fraction fraction in fractions) {
		denoms.add (fraction.denom);
		if (maxDenom == null || fraction.denom > maxDenom) 
			maxDenom = fraction.denom;
	}

	// keep adding the denominator until all values are integers
	int result = maxDenom;
	while (denoms.any((int denom) => result % denom != 0)) {
		result += maxDenom;
	}
	return result;
}

List<int> expandFractions (List<Fraction> nullspace) {
	int lcd = lcm (nullspace);  // technically, LCD is not right, but idc
	return nullspace.map(
		(Fraction fraction) => 
			(fraction.nom * (lcd ~/ fraction.denom)).abs()
	).toList();
}

class Matrix with IterableMixin <List<double>>{
	final int rows, cols;
	final List <List<Fraction>> matrix;
	Matrix._ (this.rows, this.cols, this.matrix);

	Iterator <List<double>> get iterator => matrix.map(
		(List<Fraction> row) => row.map (
			(Fraction fraction) => fraction.toNum
		)
	).iterator;

	factory Matrix (List <List<int>> matrix) {
		final List<List<Fraction>> result = [];
		int cols, rows = 0;
		for (final List<int> row in matrix) {
			final List<Fraction> rowResult = [];
			if (cols == null) cols = row.length;
			else if (row.length != cols) throw "Inconsistent Dimensions";
			for (final int index in range (row.length))
				rowResult.add (Fraction (row [index], 1));
			
			rows++;
			result.add(rowResult);
		} 
		return Matrix._ (rows, cols, result);
	}

	factory Matrix.fromDimensions(
		rows, 
		cols, 
		List<Fraction> list
	) {
		final List <List<Fraction>> matrix = [];
		int index = 0;

		for (final int _ in range (rows)) {
			final List<Fraction> row = [];
			for (final int __ in range (cols))
				row.add (list [index++]);

			matrix.add(row);
		}

		if (index != list.length) throw IndexError(
			rows * cols,  // what was invalid
			matrix,  // what is being indexed
			"rows, cols",  // the invalid arguments
			"Invalid dimensions for matrix",  // Error message
			index  // current index
		);

		return Matrix._ (rows, cols, matrix);
	}


	Fraction operator [] (Slice slice) => matrix [slice.start] [slice.end];
	operator == (dynamic other) => other is Matrix && other.matrix == matrix;
	int get hashCode => matrix.hashCode;
	int get length => rows * cols;

	String toString() => 
		"Matrix(" + 
		matrix.map (
			(List<Fraction> row) => 
				"\n\t" + row.map ((Fraction num) => num.toString()).join(", ")
		).join ("") + 
		"\n)";

	List<double> get flatten {
		final List<double> result = [];
		for (List<double> row in this) {
			for (double num in row) result.add (num);
		}
		return result;
	}

	static Pivot getPivot(List<double> col) {
		Pair result;

		for (final Pair<double> pair in enumerate<double> (col)) {
			if (pair.value != 0) return Pivot (pair);
			else result = pair;
		}

		if (col.every ((double num) => num == 0)) return Pivot.empty();
		else return Pivot (result);
	}

	Matrix get rref {
		final List<double> matrix = this.flatten;

		List<double> getCol (int index) => matrix.getRange (index, cols);

		void swapRows (int a, int b) {
			final int start1 = a * cols;
			final int end1 = (a + 1) * cols;
			final int start2 = b * cols;
			final int end2 = (b + 1) * cols;

			final List<double> temp = matrix.getRange(start1, end1);
			matrix.replaceRange(start1, end1, matrix.getRange (start2, end2));
			matrix.replaceRange(start2, end2, temp);
		}

		void crossCancel (
			int row, 
			double value, 
			int pivotRow, 
			double pivotValue
		) {
			final int offset = (pivotRow - row) * cols;
			for (final int n in range ((row + 1) * cols, start: row * cols))
				matrix [n] = (
					pivotValue * matrix [n] - value * matrix [n + offset]
				);
		}

		int pivotRow = 0, pivotCol = 0;
		final List<int> pivots = [];

		while (pivotCol < cols && pivotRow < rows) {
			final List<double> col = getCol (pivotCol);
			final Pivot pivot = getPivot (col.getRange (pivotRow, col.length));

			if (pivot.index == null) {
				pivotCol++;
				continue;
			}

			pivots.add(pivotCol);
			if (pivot.index != 0) swapRows (pivotRow, pivot.index + pivotRow);

			for (final int row in range (rows)) {
				if (row == pivotRow) continue;

				final double value = matrix [row * cols + pivotCol];
				if (value == 0) continue;
				else crossCancel (row, value, pivotRow, pivot.value);
			}
			pivotRow++;
		}

		List<Fraction> result = matrix.map (
			(double value) => Fraction (value, 1)
		).toList();

		for (Pair<int> pair in enumerate (pivots)) {
			final int temp = pair.index * cols + pair.value;
			final double value = matrix [temp];
			result [temp] = Fraction (1, 1);


			for (final int index2 in range (
				(pair.index + 1) * cols, 
				start: temp + 1)
			) result [index2] = Fraction (matrix [index2], value);
		}

		return Matrix.fromDimensions(rows, cols, result);
	}

	// List<int> get nullspace {
	// 	Matrix rref = this.rref;
	// 	List<Fraction> result = [];
	// 	for (Pair<Fraction> pair in enumerate (this)) {

	// 	}
	// }

}

File: lib/helpers/matrix.py
from my_stuff.misc import init

def lcm (nums) -> int: 
    max_denom = float("-inf")
    denoms = set()
    for fraction in nums: 
        denoms.add(fraction.denom)
        if fraction.denom > max_denom: max_denom = fraction.denom

    result = max_denom
    while any (result % denom for denom in denoms): 
        result += max_denom

    return result

def expand_fractions(nullspace: [Fraction]) -> [int]: 
    lcd = lcm (nullspace)
    return [
        abs (val.num * (lcd // val.denom))
        for val in nullspace
    ]


class Fraction: 
    @init
    def init (self, num, denom, negative): pass
    def __repr__(self): return f"{'-' if self.negative else ''}({self.num}/{self.denom})"
    def __new__(cls, num, denom): 
        if denom == 0: raise ZeroDivisionError()
        negative: bool = num >= 0 != denom >= 0
        num = abs (num)
        denom = abs (denom)
        for n in range (num, 1, -1):  # simplify
            if not num % n and not denom % n: 
                num //= n
                denom //= n

        fraction: Fraction = object.__new__(cls)
        fraction.init(num, denom, negative)
        return fraction


class Matrix: 
    @init
    def __init__(self, matrix): self.rows, self.cols = self.set_shape()
    def __getitem__(self, p): return self.matrix [p[0]] [p[1]]
    def __eq__(self, other): return self.matrix == other.matrix
    def __len__(self): return self.rows * self.cols
    def __iter__(self): return iter (self.matrix)
    def __str__(self): return repr (self)
    def __repr__(self): return (
        "Matrix(" + 
        "".join (
            [
                f"\n\t{', '.join (map (str, row))}"
                for row in self
            ]
        ) + 
        "\n)"
    )

    def flatten(self): return [num for row in self for num in row]

    def set_shape(self): 
        rows = 0
        cols = None
        for row in self.matrix:
            if cols is None: cols = len (row)
            elif len (row) != cols: raise SyntaxError("Inconsistent dimensions")
            rows += 1
        return rows, cols

    def fromDimensions(rows, cols, List):
        result = []
        index = 0
        for row in range (rows):
            values = []
            for col in range (cols): 
                values.append (List [index])
                index += 1
            result.append (values)
        if index != len (List): raise SyntaxError("fromDimensions failed")
        return Matrix (result)

    def get_pivot (self, col: list) -> (int, int): 
        result_index = None
        result = None
        for index, value in enumerate (col): 
            if value != 0: return index, value
            else: 
                result_index = index
                result = value

        else:
            if all (num == 0 for num in col): return None, None
            else: return result_index, result

    def rref(self):
        cols = self.cols
        matrix = self.flatten()

        def get_col (index): return matrix [index::cols]

        def swap_rows (row1, row2): 
            index1_1 = row1 * cols
            index1_2 = (row1 + 1) * cols
            index2_1 = row2 * cols
            index2_2 = (row2 + 1) * cols
            matrix [index1_1 : index1_2], matrix [index2_1 : index2_2] = (
                matrix [index2_1 : index2_2], matrix [index1_1 : index1_2]
            )

        def cross_cancel (row, value, pivot_row, pivot_value): 
            offset = (pivot_row - row) * cols
            for n in range (row * cols, (row + 1) * cols): 
                matrix [n] = pivot_value * matrix [n] - value * matrix [n + offset]

        pivot_row = 0
        pivot_col = 0
        pivots = []
        while pivot_col < cols and pivot_row < self.rows:
            offset, value = self.get_pivot (get_col (pivot_col) [pivot_row:])

            if offset is None: 
                pivot_col += 1
                continue

            pivots.append (pivot_col)
            if offset != 0: swap_rows (pivot_row, offset + pivot_row)

            for row in range (self.rows): 
                if row == pivot_row: continue

                val = matrix [row * cols + pivot_col]
                if val == 0: continue
                else: cross_cancel (row, val, pivot_row, value)

            pivot_row += 1

        for index, col in enumerate (pivots):
            temp = index * cols + col
            value = matrix [temp]
            matrix [temp] = 1
            for index2 in range (temp + 1, (index + 1) * cols): 
                matrix [index2] = Fraction (matrix [index2], value)

        return Matrix.fromDimensions(self.rows, self.cols, matrix)

    def nullspace(self): 
        rref = self.rref()
        nullspace = [rref [n, -1] for n in range (rref.rows)]
        nullspace.extend ([1 for _ in range (rref.cols - len (nullspace))])

        for index, value in enumerate (nullspace): 
            if value == 0: nullspace [index] = 1

        return expand_fractions(nullspace)
